### PR TITLE
Add Spack installation resource

### PR DESCRIPTION
**Pull request by rkilpadi**
Description of changes
- Create a basic resource to install Spack in the default cluster user's home directory.
- Add cluster dependent Spack packages.yaml files and configure Spack to find compilers and external directories
- Add binaries to Spack if they are successfully verified

Tests
Inspec tests to ensure that Spack directories have been created correctly and that user can successfully run spack find and spack install.
EC2 ./kitchen.ec2.sh platform-install verify spack -c 5

### CODE REVIEW

File: cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+# limitations under the License.
+
+unified_mode true
+default_action :install_spack
+
+property :spack_user, String, required: false,
+         default: node['cluster']['cluster_user']
+
+property :spack_root, String, required: false,
+         default: "/home/#{node['cluster']['cluster_user']}/spack"
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:21]: avoid hardcoding "/home" here

File: cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    source 'spack/modules.yaml'
+    owner new_resource.spack_user
+    group new_resource.spack_user
+    mode '0755'
+  end
+
+  bash 'setup Spack' do
+    user new_resource.spack_user
+    group new_resource.spack_user
+    environment({ "HOME" => "/home/#{new_resource.spack_user}" })
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb]: avoid hardcoding this, may need to be a property

File: cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      #{spack} compiler add --scope site
+
+      # Add external packages to Spack user's packages.yaml config
+      #{spack} external find --scope site
+
+      # Remove all autotools/buildtools packages. These versions need to be managed by spack or it will
+      # eventually end up in a version mismatch (e.g. when compiling gmp).
+      #{spack} tags build-tools | xargs -I {} #{spack} config --scope site rm packages:{}
+
+      # Add binary cache for common applications
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb]: pull this out into it's own action and potentially remove the logic below

File: cookbooks/aws-parallelcluster-platform/resources/spack/spack_amazon2.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  %w(autoconf automake bison byacc cscope ctags diffstat doxygen elfutils flex gcc gcc-c++ gcc-gfortran git
+     indent intltool libtool patch patchutils rcs rpm-build rpm-sign subversion swig system-rpm-config systemtap
+     curl findutils hostname iproute redhat-lsb-core python3 python3-setuptools unzip python-boto3)
+end
+
+def libfabric_path
+  '/opt/amazon/efa/lib64/pkgconfig/libfabric.pc'
+end
+
+action :install_spack do
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/spack_amazon2.rb]: make this a setup action

File: cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+unified_mode true
+default_action :install_spack
+
+property :spack_user, String, required: false,
+         default: node['cluster']['cluster_user']
+
+property :spack_root, String, required: false,
+         default: "/home/#{node['cluster']['cluster_user']}/spack"
+
+action :install_spack_common do
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb]: this should probably be just `install_spack`

File: cookbooks/aws-parallelcluster-platform/resources/spack/spack_ubuntu22.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    line '      f77: /usr/bin/g++'
+    replace_only true
+  end
+
+  replace_or_add 'fix gcc12 fc path' do
+    path "#{node['cluster']['spack']['root']}/etc/spack/compilers.yaml"
+    pattern '      fc: null'
+    line '      fc: /usr/bin/g++'
+    replace_only true
+  end
```
Comment (dreambeyondorange) [inline cookbooks/aws-parallelcluster-platform/resources/spack/spack_ubuntu22.rb]: Let's make encapsulate this in a separate action and call it here

File: cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb (hunk 1 above)
Comment (rkilpadi) [reply cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:21]: I see hardcoded `/home`s throughout the codebase, will make a PR to address these if I have some extra time.

**Review by dreambeyondorange — APPROVED** [review]

**Pull request merged**